src/patient-registration/input/custom-input/address/address-input.component.tsx

```diff
@@ -9,16 +9,105 @@ interface AddressInputProps {
   stateProvinceName: string;
   countryName: string;
   postalCodeName: string;
+  addressTemplate?: string;
 }
 
-export const AddressInput: React.FC<AddressInputProps> = ({
+export const AddressInput: React.FC<AddressInputProps> = props => {
+  return <AddressFieldRenderer {...props} />;
+};
+
+function AddressFieldRenderer(props: AddressInputProps) {
+  if (props.addressTemplate) {
+    return <AddressTemplateFieldRenderer addressTemplate={props.addressTemplate} />;
+  } else {
+    return <DefaultAddressFieldRenderer {...props} />;
+  }
+}
+
+function parseString(v: string) {
+  return new DOMParser().parseFromString(v, 'text/xml');
+}
+
+function AddressTemplateFieldRenderer({ addressTemplate }) {
+  function getTagAsDocument(tagName: string, template: XMLDocument) {
+    let tmp = template.getElementsByTagName(tagName)[0];
+    return tmp ? parseString(tmp.outerHTML) : parseString('');
+  }
+
+  function getFieldValue(field: string, doc: XMLDocument) {
+    let fieldElement = doc.getElementsByName(field)[0];
+    return fieldElement ? fieldElement.getAttribute('value') : '';
+  }
+
+  function isRequired(field: string, doc: XMLDocument) {
+    let fields = doc.getElementsByTagName('requiredElements')[0];
+    if (fields) {
+      let fieldsText: string[] = Array.prototype.map.call(
+        fields.getElementsByTagName('string'),
+        ({ textContent }) => textContent,
+      );
+      return fieldsText.includes(field);
+    }
+    return false;
+  }
+
+  function renderAddressFields() {
+    const templateXmlDoc = new DOMParser().parseFromString(addressTemplate, 'text/xml');
```

Comment:
Why not use `parseString` here? I thought it was defined for this use above.

Reply:
Thanks for the catch. `parseString` was defined after this was already written. While refactoring, I must have missed this.